### ISSUE TITLE
[Bug] After storing the MoE's checkpoint, the OPD will become slower in training.

### ISSUE THREAD

**Issue by Sniper970119** [THUDM/slime#1701]
### Bug Description

When I was using OPD to train the MoE, the model training speed would become extremely slow after storing the checkpoint. However, the Dense model does not behave like this.

MoE model（save at 200step)：

<img width="2209" height="640" alt="Image" src="https://github.com/user-attachments/assets/a411f3f9-4094-400f-9bb7-b35a2f1aeaa0" />

Dense model（save each 20 step)：
<img width="2212" height="659" alt="Image" src="https://github.com/user-attachments/assets/e5ee9313-d103-4c6b-8532-f31ec9fd6dc1" />

I don't know how to identify the problem. Could someone offer some suggestions?

### Steps to Reproduce

-

### Expected Behavior

-

### Actual Behavior

-

### Environment

- slime version: latest
- Python version: 3.11
- PyTorch version: -
- CUDA/ROCm version: 12.8
- GPU type and count: 32*H100
- OS: 
- SGLang version (if relevant):
- Megatron-LM version (if relevant):

### Logs

```shell

```

### Pre-submission Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/THUDM/slime/blob/main/CONTRIBUTING.md) and understand the collaboration scope.
- [x] I have read the [documentation](https://thudm.github.io/slime/) and my issue is not addressed there.
- [x] I have searched for [existing issues](https://github.com/THUDM/slime/issues) and this is not a duplicate.
- [x] I have provided a minimal, reproducible example.

**Comment by Sniper970119** [comment]
<img width="834" height="380" alt="Image" src="https://github.com/user-attachments/assets/a83e7525-171e-463a-95b1-9714a5b76546" />

**Comment by passionworkeer** [comment]
This is likely a checkpoint loading issue specific to MoE. Debug suggestions: 1) Check if 	orch.load is using mmap for memory efficiency, 2) Verify expert weights are properly sharded after loading, 3) Add 	orch.cuda.synchronize() before timing to get accurate measurements, 4) Check if checkpoint includes unnecessary state (like optimizer stats). The large H100 cluster suggests this might be a distributed training issue - verify 	orch.distributed state is properly restored.

**Comment by Sniper970119** [comment]
> This is likely a checkpoint loading issue specific to MoE. Debug suggestions: 1) Check if orch.load is using mmap for memory efficiency, 2) Verify expert weights are properly sharded after loading, 3) Add orch.cuda.synchronize() before timing to get accurate measurements, 4) Check if checkpoint includes unnecessary state (like optimizer stats). The large H100 cluster suggests this might be a distributed training issue - verify orch.distributed state is properly restored.

Thank you for your reply. It seems that what you mentioned are all issues related to saving. Do these affect the subsequent calculations after saving the checkpoint? If so, how should I carry out these checks?

**Comment by Sniper970119** [comment]
I directly restarted the experiment. I resumed the training of MoE from the most recent checkpoint (200 steps). It seems that this approach can restore the speed. 
I'm wondering if there was some state that didn't get restored after saving the checkpoint or something like that?

<img width="2189" height="632" alt="Image" src="https://github.com/user-attachments/assets/df631589-5b02-4f5e-aa74-9ba9d51122e9" />

**Comment by lawrence-harmonic** [comment]
I saw a similar issue, except in my case even without checkpoint save, I had slow training from the second step onwards. The issue can be traced to https://github.com/THUDM/slime/commit/796d1e4893e593675c22c8d4ce3a5c6ba01473a5, in particular if we revert as follows:
```diff
--- a/slime/backends/megatron_utils/model.py
+++ b/slime/backends/megatron_utils/model.py
@@ -28,6 +28,17 @@
 from .checkpoint import load_checkpoint, save_checkpoint
 from .data import DataIterator, get_batch
 from .loss import loss_function
+
+
+def finalize_model_grads_with_empty_cache(*args, **kwargs):
+    # Trigger empty cache when there are less than 10% free memory before the final reduce scatter.
+    device = torch.cuda.current_device()
+    free, total = torch.cuda.mem_get_info(device)
+    if free / total < 0.1:
+        clear_memory()
+    return finalize_model_grads(*args, **kwargs)
+
+
 from .model_provider import get_model_provider_func, wrap_model_provider_with_freeze
 
 logger = logging.getLogger(__name__)
@@ -536,7 +547,7 @@ def train(
         config.param_sync_func = [model_chunk.start_param_sync for model_chunk in model]
         if len(model) == 1:
             config.param_sync_func = config.param_sync_func[0]
-    config.finalize_model_grads_func = finalize_model_grads
+    config.finalize_model_grads_func = finalize_model_grads_with_empty_cache
 
     pre_hook_enabled = False
```
then the issue goes away. I would guess that the issue is not specifically checkpoint save, but anything that takes memory, such as weight sync.

**Comment by Sniper970119** [comment]
I noticed that there is a memory overflow Todo in the method's comment. That is to say, the reason for the slower subsequent training step is that the  saving of the checkpoint (or other operations) , and caused a memory overflow? thereby slowing down the subsequent training process?

> I saw a similar issue, except in my case even without checkpoint save, I had slow training from the second step onwards. The issue can be traced to [796d1e4](https://github.com/THUDM/slime/commit/796d1e4893e593675c22c8d4ce3a5c6ba01473a5), in particular if we revert as follows:
> 
> --- a/slime/backends/megatron_utils/model.py
> +++ b/slime/backends/megatron_utils/model.py
> @@ -28,6 +28,17 @@
>  from .checkpoint import load_checkpoint, save_checkpoint
>  from .data import DataIterator, get_batch
>  from .loss import loss_function
> +
> +
> +def finalize_model_grads_with_empty_cache(*args, **kwargs):
> +    # Trigger empty cache when there are less than 10% free memory before the final reduce scatter.
> +    device = torch.cuda.current_device()
> +    free, total = torch.cuda.mem_get_info(device)
> +    if free / total < 0.1:
> +        clear_memory()
> +    return finalize_model_grads(*args, **kwargs)
> +
> +
>  from .model_provider import get_model_provider_func, wrap_model_provider_with_freeze
>  
>  logger = logging.getLogger(__name__)
> @@ -536,7 +547,7 @@ def train(
>          config.param_sync_func = [model_chunk.start_param_sync for model_chunk in model]
>          if len(model) == 1:
>              config.param_sync_func = config.param_sync_func[0]
> -    config.finalize_model_grads_func = finalize_model_grads
> +    config.finalize_model_grads_func = finalize_model_grads_with_empty_cache
>  
>      pre_hook_enabled = False
> then the issue goes away. I would guess that the issue is not specifically checkpoint save, but anything that takes memory, such as weight sync.
